### DISCUSSION TITLE
Add truly minimal rendering

### DISCUSSION THREAD

**Discussion by JosiahParry** [quarto-dev/quarto-cli#7987] [Feature Requests]
I would like to be able to render quarto docs to truly minimal html but am only able to get part of the way there with the following front-matter

```yaml
format:
   html:
      minimal: true
      highlight-style: none
      self-contained: true
```

However, this still add a lot of style tags with css. Is there a way to get truly minimal rendering? e.g. pure html without the style? The idea being that I can use the rendered html in a different site perhaps one that even doesn't use bootstrap! :)

**Comment by cscheid** [comment]
@JosiahParry (tagging you here just to make sure you see that we've moved this into a discussion)

Which tags would you prefer not to see? It might be possible for you to write a post-processing script in Deno that strips tags using Deno's DOM API.

Just to temper expectations, I don't think this is very practical for us to implement, because a lot of our features use CSS for semantic annotations. For example, different floats have different classes in their captions, so they can be potentially styled differently. Multiple-column layouts are also annotated in CSS, etc.

We can't _not_ emit those, and it would be a large engineering effort to support disabling parts of the HTML format processing at this fine-grained level.

**Reply by JosiahParry** [reply]
Specifically, its the `<style>` tags in the header and use of inline style. For example using the above minimal output the below is included in the `<head>` tag. 

```
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
width: 0.8em;
margin: 0 0.8em 0.2em -1em;  vertical-align: middle;
}
.display.math{display: block; text-align: center; margin: 0.5rem auto;}
</style>


<style type="text/css">.top-right{position:absolute;top:1em;right:1em}.hidden{display:none !important}.zindex-bottom{z-index:-1 !important}figure.figure{display:block}.quarto-layout-panel{margin-bottom:1em}.quarto-layout-panel>figure{width:100%}.quarto-layout-panel>figure>figcaption,.quarto-layout-panel>.panel-caption{margin-top:10pt}.quarto-layout-panel>.table-caption{margin-top:0px}.table-caption p{margin-bottom:.5em}.quarto-layout-row{display:flex;flex-direction:row;align-items:flex-start}.quarto-layout-valign-top{align-items:flex-start}.quarto-layout-valign-bottom{align-items:flex-end}.quarto-layout-valign-center{align-items:center}.quarto-layout-cell{position:relative;margin-right:20px}.quarto-layout-cell:last-child{margin-right:0}.quarto-layout-cell figure,.quarto-layout-cell>p{margin:.2em}.quarto-layout-cell img{max-width:100%}.quarto-layout-cell .html-widget{width:100% !important}.quarto-layout-cell div figure p{margin:0}.quarto-layout-cell figure{display:block;margin-inline-start:0;margin-inline-end:0}.quarto-layout-cell table{display:inline-table}.quarto-layout-cell-subref figcaption,figure .quarto-layout-row figure figcaption{text-align:center;font-style:italic}.quarto-figure{position:relative;margin-bottom:1em}.quarto-figure>figure{width:100%;margin-bottom:0}.quarto-figure-left>figure>p,.quarto-figure-left>figure>div{text-align:left}.quarto-figure-center>figure>p,.quarto-figure-center>figure>div{text-align:center}.quarto-figure-right>figure>p,.quarto-figure-right>figure>div{text-align:right}.quarto-figure>figure>div.cell-annotation,.quarto-figure>figure>div code{text-align:left}figure>p:empty{display:none}figure>p:first-child{margin-top:0;margin-bottom:0}figure>figcaption{margin-top:.5em}figure.quarto-float-lst>figcaption{margin-bottom:.5em}div[id^=tbl-]{position:relative}.quarto-figure>.anchorjs-link{position:absolute;top:.6em;right:.5em}div[id^=tbl-]>.anchorjs-link{position:absolute;top:.7em;right:.3em}.quarto-figure:hover>.anchorjs-link,div[id^=tbl-]:hover>.anchorjs-link,h2:hover>.anchorjs-link,h3:hover>.anchorjs-link,h4:hover>.anchorjs-link,h5:hover>.anchorjs-link,h6:hover>.anchorjs-link,.reveal-anchorjs-link>.anchorjs-link{opacity:1}#title-block-header{margin-block-end:1rem;position:relative;margin-top:-1px}#title-block-header .abstract{margin-block-start:1rem}#title-block-header .abstract .abstract-title{font-weight:600}#title-block-header a{text-decoration:none}#title-block-header .author,#title-block-header .date,#title-block-header .doi{margin-block-end:.2rem}#title-block-header .quarto-title-block>div{display:flex}#title-block-header .quarto-title-block>div>h1{flex-grow:1}#title-block-header .quarto-title-block>div>button{flex-shrink:0;height:2.25rem;margin-top:0}tr.header>th>p:last-of-type{margin-bottom:0px}table,.table{caption-side:top;margin-bottom:1.5rem}figure.quarto-float-tbl figcaption,caption,.table-caption{padding-top:.5rem;padding-bottom:.5rem;text-align:center}.utterances{max-width:none;margin-left:-8px}iframe{margin-bottom:1em}details{margin-bottom:1em}details[show]{margin-bottom:0}details>summary{color:#6f6f6f}details>summary>p:only-child{display:inline}pre.sourceCode,code.sourceCode{position:relative}p code:not(.sourceCode){white-space:pre-wrap}code{white-space:pre}@media print{code{white-space:pre-wrap}}pre>code{display:block}pre>code.sourceCode{white-space:pre}pre>code.sourceCode>span>a:first-child::before{text-decoration:none}pre.code-overflow-wrap>code.sourceCode{white-space:pre-wrap}pre.code-overflow-scroll>code.sourceCode{white-space:pre}code a:any-link{color:inherit;text-decoration:none}code a:hover{color:inherit;text-decoration:underline}ul.task-list{padding-left:1em}[data-tippy-root]{display:inline-block}.tippy-content .footnote-back{display:none}.tippy-content{overflow-x:scroll}.quarto-embedded-source-code{display:none}.quarto-unresolved-ref{font-weight:600}.quarto-cover-image{max-width:35%;float:right;margin-left:30px}.cell-output-display .widget-subarea{margin-bottom:1em}.cell-output-display:not(.no-overflow-x),.knitsql-table:not(.no-overflow-x){overflow-x:auto}.panel-input{margin-bottom:1em}.panel-input>div,.panel-input>div>div{display:inline-block;vertical-align:top;padding-right:12px}.panel-input>p:last-child{margin-bottom:0}.layout-sidebar{margin-bottom:1em}.layout-sidebar .tab-content{border:none}.tab-content>.page-columns.active{display:grid}div.sourceCode>iframe{width:100%;height:300px;margin-bottom:-0.5em}a{text-underline-offset:3px}div.ansi-escaped-output{font-family:monospace;display:block}.ansi-black-fg{color:#3e424d}.ansi-black-bg{background-color:#3e424d}.ansi-black-intense-fg{color:#282c36}.ansi-black-intense-bg{background-color:#282c36}.ansi-red-fg{color:#e75c58}.ansi-red-bg{background-color:#e75c58}.ansi-red-intense-fg{color:#b22b31}.ansi-red-intense-bg{background-color:#b22b31}.ansi-green-fg{color:#00a250}.ansi-green-bg{background-color:#00a250}.ansi-green-intense-fg{color:#007427}.ansi-green-intense-bg{background-color:#007427}.ansi-yellow-fg{color:#ddb62b}.ansi-yellow-bg{background-color:#ddb62b}.ansi-yellow-intense-fg{color:#b27d12}.ansi-yellow-intense-bg{background-color:#b27d12}.ansi-blue-fg{color:#208ffb}.ansi-blue-bg{background-color:#208ffb}.ansi-blue-intense-fg{color:#0065ca}.ansi-blue-intense-bg{background-color:#0065ca}.ansi-magenta-fg{color:#d160c4}.ansi-magenta-bg{background-color:#d160c4}.ansi-magenta-intense-fg{color:#a03196}.ansi-magenta-intense-bg{background-color:#a03196}.ansi-cyan-fg{color:#60c6c8}.ansi-cyan-bg{background-color:#60c6c8}.ansi-cyan-intense-fg{color:#258f8f}.ansi-cyan-intense-bg{background-color:#258f8f}.ansi-white-fg{color:#c5c1b4}.ansi-white-bg{background-color:#c5c1b4}.ansi-white-intense-fg{color:#a1a6b2}.ansi-white-intense-bg{background-color:#a1a6b2}.ansi-default-inverse-fg{color:#fff}.ansi-default-inverse-bg{background-color:#000}.ansi-bold{font-weight:bold}.ansi-underline{text-decoration:underline}:root{--quarto-body-bg: #fff;--quarto-body-color: #222;--quarto-text-muted: #6f6f6f;--quarto-border-color: #6f6f6f;--quarto-border-width: 1px;--quarto-border-radius: 4px}table.gt_table{color:var(--quarto-body-color);font-size:1em;width:100%;background-color:rgba(0,0,0,0);border-top-width:inherit;border-bottom-width:inherit;border-color:var(--quarto-border-color)}table.gt_table th.gt_column_spanner_outer{color:var(--quarto-body-color);background-color:rgba(0,0,0,0);border-top-width:inherit;border-bottom-width:inherit;border-color:var(--quarto-border-color)}table.gt_table th.gt_col_heading{color:var(--quarto-body-color);font-weight:bold;background-color:rgba(0,0,0,0)}table.gt_table thead.gt_col_headings{border-bottom:1px solid currentColor;border-top-width:inherit;border-top-color:var(--quarto-border-color)}table.gt_table thead.gt_col_headings:not(:first-child){border-top-width:1px;border-top-color:var(--quarto-border-color)}table.gt_table td.gt_row{border-bottom-width:1px;border-bottom-color:var(--quarto-border-color);border-top-width:0px}table.gt_table tbody.gt_table_body{border-top-width:1px;border-bottom-width:1px;border-bottom-color:var(--quarto-border-color);border-top-color:currentColor}div.columns{display:initial;gap:initial}div.column{display:inline-block;overflow-x:initial;vertical-align:top;width:50%}.code-annotation-tip-content{word-wrap:break-word}.code-annotation-container-hidden{display:none !important}dl.code-annotation-container-grid{display:grid;grid-template-columns:min-content auto}dl.code-annotation-container-grid dt{grid-column:1}dl.code-annotation-container-grid dd{grid-column:2}pre.sourceCode.code-annotation-code{padding-right:0}code.sourceCode .code-annotation-anchor{z-index:100;position:absolute;right:.5em;left:inherit;background-color:rgba(0,0,0,0)}input[type=checkbox]{margin-right:.5ch}:root{--mermaid-bg-color: #fff;--mermaid-edge-color: #999;--mermaid-node-fg-color: #222;--mermaid-fg-color: #222;--mermaid-fg-color--lighter: #3c3c3c;--mermaid-fg-color--lightest: #555555;--mermaid-font-family: sans-serif;--mermaid-label-bg-color: #fff;--mermaid-label-fg-color: #468;--mermaid-node-bg-color: rgba(68, 102, 136, 0.1);--mermaid-node-fg-color: #222}@media print{:root{font-size:11pt}#quarto-sidebar,#TOC,.nav-page{display:none}.page-columns .content{grid-column-start:page-start}.fixed-top{position:relative}.panel-caption,.figure-caption,figcaption{color:#666}}
</style>
```

**Reply by JosiahParry** [reply]
^ if you line wrap, that last style tag is quite the bad offender.

**Reply by JosiahParry** [reply]
Really, my hope is that by setting `minimal: true` i  would get minimal html similar to what I might get by just using pandoc. the problem with "just using pandoc", though, is that the quarto divs and spans are not recognized.

**Reply by JosiahParry** [reply]
This is related to https://github.com/quarto-dev/quarto-cli/issues/3933
As stated in the issue: 

>  "I wonder if minimal: true is enough. I think we need theme: none also to avoid adding some CSS in the document."

In my case `minimal: true` is not enough

**Reply by cscheid** [reply]
> ^ if you line wrap, that last style tag is quite the bad offender.

I hear you. But the issue is that it contains rules like this: `.quarto-layout-cell .html-widget ...`

Like I mentioned in my first comment, this rule and others are necessary for Quarto HTML to work, independently of Bootstrap vs. not. We need them in the "minimal" version of the document.

I really think you're best served by a [post-render Deno script](https://quarto.org/docs/projects/scripts.html#pre-and-post-render).

**Reply by cscheid** [reply]
> the problem with "just using pandoc", though, is that the quarto divs and spans are not recognized.

Can you give an example? We might be able to help. It might be the case, too, that your workflow works best if you use quarto to generate markdown output, and then use Pandoc to turn that into "truly minimal HTML".

**Reply by JosiahParry** [reply]
Thanks for the response! 

It seems that my issue is very closely related to https://github.com/quarto-dev/quarto-cli/issues/3933. I see that it is closed though its unclear what the resolution is. Will this resolve me issue in a future release? 

Post-rendering seems like it may be useful however, the process is not documented. The examples all use some variation of `quit()` for pre-rendering. Nor does it mention what is possible or how it might be done.

**Reply by JosiahParry** [reply]
I've got my dirty work around using 

```yaml
format:
   html:
      minimal: true
      highlight-style: none
      exmbed-resources: true
``` 

with an R script using `rvest` and `xml2` for post-processing afterwards (though it may be nice to use that in the post-render section) so I'll close this. Thanks again!

**Reply by cscheid** [reply]
> It seems that my issue is very closely related to https://github.com/quarto-dev/quarto-cli/issues/3933. I see that it is closed though its unclear what the resolution is.

The resolution is pointed to by [the issue-closing commit](https://github.com/quarto-dev/quarto-cli/commit/2db5ad3bd1af6b34e3e527e84aa28d3d24d4dddf). We use this a lot in Quarto, so here it is for your future reference:

<img width="809" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/de900b93-2321-4bb8-bc8c-45b885f48567">

The `2db5ad3...` link points to the commit and the message explaining it.

> Post-rendering seems like it may be useful however, the process is not documented. [...] Nor does it mention what is possible or how it might be done.

You're right that we should document example scenarios. But this is a general Deno (or Python, or R, if you're more familiar with those) script; you can do anything at all that you want. The only thing that happens is that it runs after a project rendering is finished.

**Reply by cscheid** [reply]
> with an R script using rvest and xml2 for post-processing afterwards (though it may be nice to use that in the post-render section) so I'll close this. 

Ok, happy to hear! You did exactly what I would in this case, except for precisely what you said, which is that you can run that script automatically after each rendering by making it a post-render script.

**Reply by JosiahParry** [reply]
I was able to read the commit message but not the typescript 😅 so back at square one! 

I understand that I can do "anything at all that [I] want" but it's not clear where to start. How do I access the result of the render? What format am I interacting with it in? Does this create some state for me to work from? Do I need to read in the rendered html and process that? Those are just some of the starting questions I have. I like to think I'm not a complete novice! But its also hard to get started when the docs don't necessarily tell you where or how to begin

**Reply by cscheid** [reply]
> I understand that I can do "anything at all that [I] want" but it's not clear where to start. How do I access the result of the render? What format am I interacting with it in? Does this create some state for me to work from? Do I need to read in the rendered html and process that? Those are just some of the starting questions I have. I like to think I'm not a complete novice! But its also hard to get started when the docs don't necessarily tell you where or how to begin

This _is_ explained in the page. We communicate with the script via environment variables: https://quarto.org/docs/projects/scripts.html#pre-and-post-render

You get that information: the project output directory, the input files, and the output files. What you do with that is entirely up to you.

**Reply by cderv** [reply]
To add some clarification about our options: 

`minimal: true` is about Quarto features. (https://quarto.org/docs/output-formats/html-basics.html#minimal-html) It should deactivate Quarto related features. 

Setting `highlight-style: none` is needed if you don't want Pandoc's feature. Pandoc will do syntax highlighting by default, and if you don't want the Pandoc's CSS you need to opt out for sure. Quarto Style has been opt out through `minimal` already. 

`theme: none` is now set by `minimal: true` I believe (following #3933), so no Pandoc HTML theme should be in the output document. 

What you see left at https://github.com/quarto-dev/quarto-cli/discussions/7987#discussioncomment-7908863 is just internal HTML set in the Pandoc's templates. it is in the style template 
https://github.com/quarto-dev/quarto-cli/blob/6d5384b5312087e7a3df9c5c81d25852eda88fb7/src/resources/formats/html/pandoc/html.styles#L178-L197

and inserted unconditionnaly for some, or based on Pandoc's feature for other

It is used in the main template. 
https://github.com/quarto-dev/quarto-cli/blob/6d5384b5312087e7a3df9c5c81d25852eda88fb7/src/resources/formats/html/pandoc/template.html#L8-L10

So if you want to remove that part, you need to tweak the templates (or the partial `styles.html()`). 

`minimal: true` alone is not expected to opt out Pandoc's own feature. 

I hope this clarifies how this works, and why you see what you see.

**Comment by mcanouil** [comment]
@JosiahParry a side note: `self-contained` is only maintain for retro-compatibility with old documents (before Quarto 1.1).
The correct option is `embed-resources` as documented in Quarto documentation.
`self-contained` was completely removed from Pandoc more than a year ago.
If you could edit your code to avoid future readers to use this option and think it's undocumented.
